### PR TITLE
Add a simple script to sender to the APEL REST interface

### PR DESCRIPTION
**Pull request by gregcorbett**
A standalone send via REST script based off the prototype SSM REST methods.

Using this replaces the need for using the SSM prototype in Indigo.

### CODE REVIEW

**Comment by gregcorbett** [comment]
Old-style class changed to fix codacy

File: scripts/sender.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+                    return response
+                else:
+                    log.warning("Could not connect to endpoint, retrying")
+                    time.sleep(attempt_number)
+                    continue
+
+            except socket.gaierror as e:
+                log.info('socket.gaierror: %s, %s',
+                         e.errno, e.strerror)
+                sys.exit()
```
Comment (tofu-rocketry) [inline scripts/sender.py]: This `sys.exit()` and the one below only happen when the script fails, right? If so, the exit code should be set to 1 for consistency and clarity.

File: scripts/sender.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+            self._outq.remove(msgid)
+
+        log.info('Tidying message directory.')
+        try:
+            # Remove empty dirs and unlock msgs older than 5 min (default)
+            self._outq.purge()
+        except OSError, e:
+            log.warn('OSError raised while purging message queue: %s', e)
+
+    def _rest_connect(self, verb, url, data, headers, expected_response_code):
```
Comment (tofu-rocketry) [inline scripts/sender.py]: Method is a little misnamed as it doesn't merely connect, but more importantly sends.

File: scripts/sender.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        """Send all the messages in the outgoing queue via REST."""
+        log.info('Found %s messages.', self._outq.count())
+        for msgid in self._outq:
+            if not self._outq.lock(msgid):
+                log.warn('Message was locked. %s will not be sent.', msgid)
+                continue
+
+            text = self._outq.get(msgid)
+
+            url = '/api/v1/cloud/record'
```
Comment (gregcorbett) [inline scripts/sender.py]: the variable should probably be renamed `path`.

the version number can be set in the config, but the path (excluding version) will always be the same.

Comment (gregcorbett) [inline scripts/sender.py]: well, it will be so long as we are only doing cloud accounting

File: scripts/sender.py (hunk 2 above)
Comment (gregcorbett) [inline scripts/sender.py]: done

File: scripts/sender.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        """
+        Send a HTTPS request to self._dest.
+
+        Will attempt to repeat if expected_response is not returned.
+        """
+        attempt_number = 0
+
+        while attempt_number < 3:
+            try:
+                attempt_number = attempt_number + 1
```
Comment (gregcorbett) [inline scripts/sender.py]: done

File: scripts/sender.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        log.info('Tidying message directory.')
+        try:
+            # Remove empty dirs and unlock msgs older than 5 min (default)
+            self._outq.purge()
+        except OSError, e:
+            log.warn('OSError raised while purging message queue: %s', e)
+
+    def _rest_connect(self, verb, url, data, headers, expected_response_code):
+        """
+        Send a HTTPS request to self._dest.
```
Comment (gregcorbett) [inline scripts/sender.py]: done

File: scripts/sender.py (hunk 1 above)
Comment (gregcorbett) [inline scripts/sender.py]: yes, changed

File: scripts/sender.py [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+                 response.status, response.reason)
+        sys.exit()
+
+
+def main():
+    """Configure a Sender and send Records to an APEL REST interface."""
+    # The web address of the accounting server you wish to send to
+    # i.e accounting.indigio-datacloud.eu
+    dest = ''
+    # The directory to read saved accounting messages to
```
Comment (gregcorbett) [inline scripts/sender.py]: "read from", done

File: scripts/sender.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    # i.e /var/spool/apel/cloud
+    qpath = ''
+    # The location of this machines private key
```
Comment (gregcorbett) [inline scripts/sender.py]: done

File: scripts/sender.py [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    # i.e. /etc/httpd/ssl/key.file
+    key = ''
+    # The location of this machines X.509 certifcate
```
Comment (gregcorbett) [inline scripts/sender.py]: done

File: scripts/sender.py [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        # if here, attempt_number has been exceeded
+        log.info('Could not connect to endpoint. Error: %s - %s',
+                 response.status, response.reason)
+        sys.exit()
+
+
+def main():
+    """Configure a Sender and send Records to an APEL REST interface."""
+    # The web address of the accounting server you wish to send to
+    # i.e accounting.indigio-datacloud.eu
```
Comment (gregcorbett) [inline scripts/sender.py]: 'e.g.', done

File: scripts/sender.py [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+                sys.exit()
+
+        # if here, attempt_number has been exceeded
+        log.info('Could not connect to endpoint. Error: %s - %s',
+                 response.status, response.reason)
+        sys.exit()
+
+
+def main():
+    """Configure a Sender and send Records to an APEL REST interface."""
```
Comment (gregcorbett) [inline scripts/sender.py]: done

**Comment by gregcorbett** [comment]
I think that's most of the changes made, except for the first one (but I have commented on it and made changes 'around' it)

File: scripts/sender.py (hunk 3 above)
Comment (tofu-rocketry) [inline scripts/sender.py]: That all sounds reasonable.

Comment (gregcorbett) [inline scripts/sender.py]: done

File: scripts/sender.py [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+
+            except socket.gaierror as e:
+                log.info('socket.gaierror: %s, %s',
+                         e.errno, e.strerror)
+                sys.exit(1)
+
+        # if here, attempt_number has been exceeded
+        log.info('Could not connect to endpoint. Error: %s - %s',
+                 response.status, response.reason)
+        sys.exit()
```
Comment (tofu-rocketry) [inline scripts/sender.py]: This exit still needs a 1, right?

File: scripts/sender.py [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+    arg_parser.add_argument("-k", "--key", type=str, required=True,
+                            help=("The path to this machine's private key. "
+                                  "e.g. /etc/httpd/ssl/key.file"))
+
+    arg_parser.add_argument("-c", "--certificate", type=str, required=True,
+                            help=("The location of this "
+                                  "machine's X.509 certifcate. "
+                                  "e.g. /etc/httpd/ssl/cert.file"))
+
+    arg_parser.add_argument("-v", "--version", type=str, required=True,
```
Comment (tofu-rocketry) [inline scripts/sender.py]: Can this be set to default to "v1"?

File: scripts/sender.py (hunk 11 above)
Comment (gregcorbett) [inline scripts/sender.py]: it has been. https://github.com/apel/rest/pull/29/commits/0a326e505fd11dad3e044a4b6f0b56ad1c2a28ef

Comment (tofu-rocketry) [inline scripts/sender.py]: That's the socket error, I mean the attempt number exit. There are two `sys.exit()`.

Comment (gregcorbett) [inline scripts/sender.py]: ah right!

Comment (gregcorbett) [inline scripts/sender.py]: done

**Comment by gregcorbett** [comment]
hmmm, I'm not sure how to test this, (in an automated way) as it needs the REST interface to respond for a meaningful test. Doesn't it?

**Comment by gregcorbett** [comment]
I'm also happy with using command line args, it takes away having to edit the python file but doesnt go the whole hog and use a config file

**Comment by tofu-rocketry** [comment]
Yeah, it is tricky to test well. You could check what happens when you give it a duff endpoint at least. It could be added to the coverage at least. I'll leave the choice up to you.

**Comment by gregcorbett** [comment]
I'm tempted to say leave it as is... or just add it to the test coverage

**Pull request merged**